### PR TITLE
test: adding e2e for grandpa client recovery test

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

This PR adds an e2e test using ibc-go-simd with 08-wasm and a polkadot/substrate counterparty for testing client recovery - i.e. an active client expires and can use a substitute client of the same type to return it to an active state.

This leverages the new contract `sudo` entrypoint `MigrateClientStore`.

closes: #4970 

---

### How to run the test

1. Create a e2e yml dedicated for 08-wasm tests. Drop it into a folder named `dev-configs` inside `e2e`

```
cd e2e
mkdir dev-configs
touch e2e-grandpa.yml
```

2. Open the file `e2e-grandpa.yml` and drop the content inside.
```
# This file contains configuration for running e2e tests.
# Many of these fields can be overridden with environment variables.
# All fields that support this have the corresponding environment variable name in a comment beside the field.

---
chains:
  # the entry at index 0 corresponds to CHAIN_A
  # NOTE: The chain A spec is overridden directly within 08-wasm e2e tests to use substrate nodes
- chainId: chainA-1
  numValidators: 2
  numFullNodes: 1
  image: ghcr.io/cosmos/ibc-go-simd # override with CHAIN_IMAGE
  tag: release-v8.0.x  # override with CHAIN_A_TAG
  binary: simd # override with CHAIN_BINARY

  # the entry at index 1 corresponds to CHAIN_B
- chainId: chainB-1
  numValidators: 1
  numFullNodes: 0
  image: ghcr.io/cosmos/ibc-go-wasm-simd # override with CHAIN_IMAGE
  tag: pr-5000  # override with CHAIN_B_TAG
  binary: simd # override with CHAIN_BINARY

activeRelayer: hyperspace # override with RELAYER_ID
relayers:
  - id: hermes
    image: ghcr.io/informalsystems/hermes
    tag: "bef2f53"
  - id: rly
    image: ghcr.io/cosmos/relayer
    tag: "latest"
  - id: hyperspace
    image: "ghcr.io/misko9/hyperspace"
    tag: "local"


cometbft:
  logLevel: info

debug:
  # setting this value to true will force log collection even if the test passes.
  dumpLogs: false

# Required only for upgrade tests.
# Chain A will be upgraded the specified tag.
# The plan name must be registered as an upgrade handler in the given tag.
upgrade:
  planName: ""
  tag: ""
```

3. Run `make e2e-test` and select the config `e2e-grandpa.yml` and search for `TestRecoverClient_Succeeds_GrandpaContract` - note the test is long running ~20 mins

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: e2e/tests/wasm/grandpa_test.go [hunk 1]
```diff
@@ -255,6 +257,160 @@ func (s *GrandpaTestSuite) TestMsgTransfer_Succeeds_GrandpaContract() {
...
+
+	pathName := s.GetPathName(0)
+
+	err = r.GeneratePath(ctx, eRep, cosmosChain.Config().ChainID, polkadotChain.Config().ChainID, pathName)
+	s.Require().NoError(err)
+
+	// create client pair with subject (bad trusting period)
+	subjectClientID := clienttypes.FormatClientIdentifier(ibcexported.Wasm, 0)
+	// TODO(damian): The hyperspace relayer makes no use of create client opts
+	// https://github.com/strangelove-ventures/interchaintest/blob/main/relayer/hyperspace/hyperspace_commander.go#L83
```
Comment (damiannolan) [inline e2e/tests/wasm/grandpa_test.go:364]: 👀

File: e2e/tests/wasm/grandpa_test.go [hunk 2]
```diff
@@ -255,6 +257,160 @@ func (s *GrandpaTestSuite) TestMsgTransfer_Succeeds_GrandpaContract() {
...
+	substituteClientID := clienttypes.FormatClientIdentifier(ibcexported.Wasm, 1)
+	s.SetupClients(ctx, r, ibc.DefaultClientOpts())
+
+	// wait for block
+	err = testutil.WaitForBlocks(ctx, 1, cosmosChain, polkadotChain)
+	s.Require().NoError(err)
+
+	// ensure subject client is expired
+	// TODO(damian): The grandpa-cw contract does not seem to have a concept of trusting period.
+	// https://github.com/ComposableFi/composable-ibc/blob/master/light-clients/ics10-grandpa-cw/src/contract.rs#L343-L357
```
Comment (damiannolan) [inline e2e/tests/wasm/grandpa_test.go:386]: 👀

File: e2e/tests/wasm/grandpa_test.go [hunk 3]
```diff
@@ -255,6 +257,160 @@ func (s *GrandpaTestSuite) TestMsgTransfer_Succeeds_GrandpaContract() {
...
+
+	polkadotChain := chainA.(*polkadot.PolkadotChain)
+	cosmosChain := chainB.(*cosmos.CosmosChain)
+
+	// we explicitly skip path creation as the contract needs to be uploaded before we can create clients.
+	r := s.ConfigureRelayer(ctx, polkadotChain, cosmosChain, nil, func(options *interchaintest.InterchainBuildOptions) {
+		options.SkipPathCreation = true
+	})
+
+	s.InitGRPCClients(cosmosChain)
```
Comment (chatton) [inline e2e/tests/wasm/grandpa_test.go:327]: I think all of this code can probably put into a setup fn for the `GrandpaTestSuite` type to be re-used in all the tests.

Comment (damiannolan) [reply e2e/tests/wasm/grandpa_test.go:327]: I 100% agree. I can do it as part of this PR, or separate it out

File: e2e/tests/data/ics10_grandpa_cw_expiry.wasm.gz
Comment (damiannolan) [inline e2e/tests/data/ics10_grandpa_cw_expiry.wasm.gz:1]: This wasm contract was compiled from this fork on @misko9's branch https://github.com/misko9/centauri/commit/44a565c1d0888a8ae0d9e1cd8d9fd5157f64f9a9 and includes an upstream change for client status expiry checks in this PR https://github.com/ComposableFi/composable-ibc/pull/448 as well as a modification to the unbonding period (1500s) and calculated trusting period (1500 /3 = 500s) to allow clients to expire faster in test scenarios.

File: e2e/tests/wasm/grandpa_test.go [hunk 4]
```diff
@@ -255,6 +257,160 @@ func (s *GrandpaTestSuite) TestMsgTransfer_Succeeds_GrandpaContract() {
...
+	status, err = s.clientStatus(ctx, cosmosChain, substituteClientID)
+	s.Require().NoError(err)
+	s.Require().Equal(ibcexported.Active.String(), status, "unexpected substitute client status")
+
+	// create and execute a client recovery proposal
+	authority, err := s.QueryModuleAccountAddress(ctx, govtypes.ModuleName, cosmosChain)
+	s.Require().NoError(err)
+	msgRecoverClient := clienttypes.NewMsgRecoverClient(authority.String(), subjectClientID, substituteClientID)
+	s.Require().NotNil(msgRecoverClient)
+	s.ExecuteAndPassGovV1Proposal(ctx, msgRecoverClient, cosmosChain, cosmosUser)
```
Comment (damiannolan) [inline e2e/tests/wasm/grandpa_test.go:401]: This test running with the contract mentioned in https://github.com/cosmos/ibc-go/pull/5027/files#r1386287352 is failing here, when the `MsgRecoverClient` is delivered by gov. It calls through ibc core and into the wasmvm returning the following:

```
client recovery failed: failed to validate substitute client: Error calling the VM: Error executing Wasm: Wasmer runtime error: RuntimeError: Error calling into the VM's backend: Panic in FFI call: wasm contract call failed: failed waiting for condition after 300.000000 seconds
```

Prior to this error being returned (before @misko9's branch was rebased with main of the upstream) I could get as far as this point: https://github.com/misko9/centauri/blob/steve/contracts/light-clients/ics10-grandpa-cw/src/contract.rs#L259 but also yielded a failure for the equality check:

```
client recovery failed: failed to validate substitute client: Grandpa error: subject client state does not match substitute client state: wasm contract call failed: failed waiting for condition after 300.000000 seconds
```

File: e2e/tests/wasm/grandpa_test.go (hunk 3 above)
Comment (damiannolan) [reply e2e/tests/wasm/grandpa_test.go]: resolved in https://github.com/cosmos/ibc-go/pull/5027/commits/dcc1a11ef6607507b39ba73ae1f3c8b2054d9515

**Comment by damiannolan** [comment]
Publishing as r4r to run CI jobs

File: e2e/tests/data/ics10_grandpa_cw_expiry.wasm.gz
Comment (damiannolan) [inline e2e/tests/data/ics10_grandpa_cw_expiry.wasm.gz:1]: I think we should move these contracts to `e2e/tests/wasm/contracts`

File: e2e/tests/wasm/grandpa_test.go [hunk 5]
```diff
@@ -373,6 +375,109 @@ func (s *GrandpaTestSuite) TestMsgMigrateContract_ContractError_GrandpaContract(
 	s.Require().ErrorContains(err, "migration not supported")
 }
 
+func (s *GrandpaTestSuite) TestRecoverClient_Succeeds_GrandpaContract() {
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply e2e/tests/wasm/grandpa_test.go:378]: Yep, was planning to add something so will do

File: e2e/tests/wasm/grandpa_test.go [hunk 6]
```diff
@@ -373,6 +375,109 @@ func (s *GrandpaTestSuite) TestMsgMigrateContract_ContractError_GrandpaContract(
...
+	// create client pair with substitue
+	substituteClientID := clienttypes.FormatClientIdentifier(ibcexported.Wasm, 1)
+	s.SetupClients(ctx, r, ibc.DefaultClientOpts())
+
+	// wait for block
+	err = testutil.WaitForBlocks(ctx, 1, cosmosChain, polkadotChain)
+	s.Require().NoError(err)
+
+	// ensure subject client is expired
+	// TODO(damian): The grandpa-cw contract does not seem to have a concept of trusting period.
```
[the remark this answers is not in the document]
Comment (damiannolan) [reply e2e/tests/wasm/grandpa_test.go:452]: I will open an issue on composable-ibc for this - but essentially trusting period is hardcoded to be calculated as `unbonding period / 3` [ref](https://github.com/ComposableFi/composable-ibc/blob/master/light-clients/common/src/lib.rs#L171-L175).

And unbonding period is decided based on an enum of the `RelayChain` embedded within grandpa client state - https://github.com/ComposableFi/composable-ibc/blob/master/light-clients/ics10-grandpa/src/proto/grandpa.proto#L50

**Comment by damiannolan** [comment]
We have a passing test run: https://github.com/cosmos/ibc-go/actions/runs/6848930361/job/18621252828?pr=5027 🎉

**Comment by damiannolan** [comment]
```
failed to store wasm bytecode: failed to store contract: Error calling the VM: Error during static Wasm validation: Wasm bytecode could not be deserialized. Deserialization error: "magic header not detected: bad magic number (at offset 0x0)": failed waiting for condition after 300.000000 seconds
```

seeing the above after gzipping everything, might just revert

**Review by crodriguezvega — APPROVED** [review]
This is just so damn good! Well done, guys!

**Review by colin-axner — APPROVED** [review]
Excellent work as always :pray: Will need a rebase before merge

File: proto/ibc/lightclients/wasm/v1/query.proto [hunk 7]
```diff
@@ -13,7 +13,7 @@ service Query {
     option (google.api.http).get = "/ibc/lightclients/wasm/v1/checksums";
   }
 
-  // Get Wasm code for given code hash
+  // Get Wasm code for given checksum
```
Comment (damiannolan) [inline proto/ibc/lightclients/wasm/v1/query.proto:16]: Looks like there's an inconsistency on main here. I re-ran `make proto-all` when resolving conflicts which produced a change in the `.pb.go` godocs with checksum -> code hash. I believe it should be "checksum", right?

**Review by chatton — APPROVED** [review]
LGTM glad we have this passing ❤️

**Pull request merged**